Step4, 5 - 자동차 경주(우승자) + (리펙토링)

Dev-RubinJo opened the pull request:
자동차 경주(우승자)를 수행했습니다. 
그와 더불어 함께 step5도 같이 수행했습니다. 확인 부탁바랍니다.

Dev-RubinJo committed:
feat(step3): 자동차 경주 미션

자동차 경주 코드를 작성했습니다.

각각의 로직에 단위테스트를 구현해봤습니다.
필요한 부분에 대하여 에러를 정의하여 처리해뒀습니다.

Dev-RubinJo committed:
feat(step3): input, output담당 struct 생성

기존 Racing과 Car에 있던 input, output 메서드를 분리하기 위해 담당 struct를 만들어서 적용했습니다.

Dev-RubinJo committed:
refactor(step3): 자동차 경주 미션 피드백 적용

1. 보다 역할에 맞도록 클래스와 메서드를 분리시킴
2. 네이밍에서 중복되는 의미 혹은 불필요한 단어가 포함된 경우에 대한 처리
3. main.swift에 레이싱 경기 시작 로직 적용
4. 메서드에서 불필요한 반환값 제거
5. 입력값 유효성 확인 로직 추가

Dev-RubinJo committed:
add(step4): 자동차 경주(우승자) README.md 추가

Dev-RubinJo committed:
feat(step4): 자동차 경주(우승자) 미션

step4의 기능 요구 사항에 따라 자동차 경주(우승자) 미션을 수행했습니다.

1. 자동차의 이름들을 입력받아 자동차 생성, ","를 기준으로 이름 구분
2. 자동차가 움직일때마다 이름과 함께 자동차 이동을 표시
3. 경주가 다 끝난 후 우승한 자동차 표시
4. Input, Output로직을 제외한 나머지 로직들 Code Coverage로 확인하며 테스트
5. indent의 depth가 2를 넘지 않도록 작성
6. 기능 명세 상 에러처리를 해야할 부분에 대해 Validator로 에러를 처리할 수 있도록 함

wonhee009 commented on RacingCar/Car.swift:
```diff
@@ -10,15 +10,23 @@ import Foundation
     
     func move(randomNumber: Int) throws {
-        if try Validator.checkValidation(number: randomNumber) >= 4 {
+        if try Validator.checkValidation(number: randomNumber) >= canMoveMinNumber {
```
`Validator.checkValidation(number: _)`가 숫자를 반환하고 있는데요.
check 혹은 validation이라는 이름이라면 반환 타입이 Boolean여야 하지 않을까 생각이 듭니다.

```swift
Validator.checkValidation(number: _) -> Bool

guard Validator.checkValidation(number: randomNumber),
           randomNumber >= canMoveMinNumber else { return }
```
이런 방식은 어떤가요

Dev-RubinJo replied:
와 제가 너무 값 체크하는 것을 이상하게 생각했네요. Validation 체크는 확실히 bool로 판단하는게 제일 명확한 것 같습니다. 적용하겠습니다.

wonhee009 commented on RacingCar/Int+Extension.swift:
```diff
@@ -0,0 +1,15 @@
+
+extension Int {
+    
+    func getBiggestInt(with car: Car) -> Int {
+        return car.moveCount > self ? car.moveCount : self
+    }
```
해당 메서드가 Int의 extension으로 정의되어야 하는 이유가 있을까요?
`Car`의 프로퍼티 중 가장 큰 moveCount를 판단하는 로직인데 모든 Int가 사용할 수 있어야 하나요?

Dev-RubinJo replied:
제가 너무 Car에서 분리해야한다는 생각만 하다가 Int extension으로 빼뒀던 것 같습니다. 다만 아직 Int의 extension으로 처리해야겠다는 생각이 있어 RacingController에서만 처리할 수 있도록 하는 방법을 생각해보겠습니다.

wonhee009 commented on RacingCar/Racing.swift:
```diff
@@ -10,14 +10,10 @@ import Foundation
-            completion(cars)
+    func run(cars: [Car], completion: ([Car]) -> Void) throws {
+        try cars.forEach { car in
+            try car.move(randomNumber: car.generateRandomNumber())
```
`car.move`를 하는데 `car.generateRandomNumber()`를 호출해서 주입해주는 부분이 어색한거 같습니다.
사실상 동일 한 car에 정의된 메서드인데요.
```swift
car { 
   func move()
   private func generateRandomNumber()
}
```
이렇게 `generateRandomNumber`을 private으로 감추고 move 로직 내에서 generateRandomNumber를 호출해도 좋을것 같구요.

랜덤한 숫자를 주입받고 싶다면 generateRandomNumber를 car가 아닌 외부로 분리해서 사용해도 좋을거 같습니다.

Dev-RubinJo replied:
처음에 작성할때는 그런 느낌을 받지 못했었는데 확실히 피드백해주시고 난 뒤 다시 보니까 car에서 랜덤 숫자 생성도 하고 그 숫자로 움직이는게 이상하긴 하군요. Generator를 만들어 처리하겠습니다!

wonhee009 commented on RacingCar/RacingCarOutput.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    private static func printMove(_ car: Car) {
+        print("\(car.name) : ", terminator: "")
+        for _ in 0 ..< car.moveCount {
+            print("-", terminator: "")
+        }
```
이런 부분들도 car 내부에 정의하는건 어떨까요?
```swift
Car {
   var moveText: String {
      ...
   }
}
```

Dev-RubinJo replied:
해당부분은 ResultView에 가깝다고 생각하여 Car에 있는 것은 부적절하다고 생각합니다.

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
코멘트 남겨두었으니 확인 부탁드려요 🙏

Dev-RubinJo committed:
fix(step4): 자동차 경주(우승자) 피드백 적용

1. Validator에서 유효성 체크하는 로직이 값을 체크하고 유효하면 그 값을 그대로 반환해주던 것에서 true / false 로 반환하도록 수정
2. Int 전체에서 getBiggestInt 메서드를 사용할 수 있던 형태에서 RacingController에서만 사용할 수있도록 수정
3. Car에서 랜덤 숫자를 생성하고 생성된 숫자를 이용하여 차량을 움직이는 로직을 Generator를 이용하여 랜덤 숫자를 생성할 수 있도록 분리

wonhee009 reviewed (approved):
수고하셨습니다 🎉
